lib/crypto/c_src/cipher.c

```diff
@@ -28,98 +28,192 @@
+
+    {{"blowfish_cbc"}, {&EVP_bf_cbc}, 0, NO_FIPS_CIPHER},
+    {{"blowfish_cfb64"}, {&EVP_bf_cfb64}, 0, NO_FIPS_CIPHER},
+    {{"blowfish_ofb64"}, {&EVP_bf_ofb}, 0, NO_FIPS_CIPHER},
+    {{"blowfish_ecb"}, {&EVP_bf_ecb}, 0, NO_FIPS_CIPHER | ECB},
+
+    {{"aes_cbc"}, {&EVP_aes_128_cbc}, 16, 0},
+    {{"aes_cbc"}, {&EVP_aes_192_cbc}, 24, 0},
+    {{"aes_cbc"}, {&EVP_aes_256_cbc}, 32, 0},
+
+    {{"aes_128_cbc"}, {&EVP_aes_128_cbc}, 16, 0},
+    {{"aes_192_cbc"}, {&EVP_aes_192_cbc}, 24, 0},
+    {{"aes_256_cbc"}, {&EVP_aes_256_cbc}, 32, 0},
+
+    {{"aes_cfb8"}, {&EVP_aes_128_cfb8}, 16, NO_FIPS_CIPHER},
+    {{"aes_cfb8"}, {&EVP_aes_192_cfb8}, 24, NO_FIPS_CIPHER},
+    {{"aes_cfb8"}, {&EVP_aes_256_cfb8}, 32, NO_FIPS_CIPHER},
+
+    {{"aes_cfb128"}, {&EVP_aes_128_cfb128}, 16, NO_FIPS_CIPHER},
+    {{"aes_cfb128"}, {&EVP_aes_192_cfb128}, 24, NO_FIPS_CIPHER},
+    {{"aes_cfb128"}, {&EVP_aes_256_cfb128}, 32, NO_FIPS_CIPHER},
+
+    {{"aes_ecb"}, {&EVP_aes_128_ecb}, 16, ECB},
+    {{"aes_ecb"}, {&EVP_aes_192_ecb}, 24, ECB},
+    {{"aes_ecb"}, {&EVP_aes_256_ecb}, 32, ECB},
+
+#if defined(HAVE_EVP_AES_CTR)
+    {{"aes_128_ctr"}, {&EVP_aes_128_ctr}, 16, 0},
+    {{"aes_192_ctr"}, {&EVP_aes_192_ctr}, 24, 0},
+    {{"aes_256_ctr"}, {&EVP_aes_256_ctr}, 32, 0},
+    {{"aes_ctr"},    {&EVP_aes_128_ctr}, 16, 0},
+    {{"aes_ctr"},    {&EVP_aes_192_ctr}, 24, 0},
+    {{"aes_ctr"},    {&EVP_aes_256_ctr}, 32, 0},
+#else
+    {{"aes_128_ctr"}, {NULL}, 16, AES_CTR_COMPAT},
+    {{"aes_192_ctr"}, {NULL}, 24, AES_CTR_COMPAT},
+    {{"aes_256_ctr"}, {NULL}, 32, AES_CTR_COMPAT},
+    {{"aes_ctr"},     {NULL}, 16, AES_CTR_COMPAT},
+#endif
+
+#if defined(HAVE_CHACHA20)
+    {{"chacha20"}, {&EVP_chacha20}, 32, NO_FIPS_CIPHER},
+#else
+    {{"chacha20"}, {NULL}, 0, NO_FIPS_CIPHER},
+#endif
+
+    /*==== AEAD ciphers ====*/
+#if defined(HAVE_CHACHA20_POLY1305)
+    {{"chacha20_poly1305"}, {&EVP_chacha20_poly1305}, 0, NO_FIPS_CIPHER | AEAD_CIPHER, {{EVP_CTRL_AEAD_SET_IVLEN,EVP_CTRL_AEAD_GET_TAG,EVP_CTRL_AEAD_SET_TAG}}},
+#else
+    {{"chacha20_poly1305"}, {NULL}, 0, NO_FIPS_CIPHER | AEAD_CIPHER, {{0,0,0}}},
+#endif
+
+#if defined(HAVE_GCM)
+    {{"aes_gcm"},     {&EVP_aes_128_gcm}, 16, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+    {{"aes_gcm"},     {&EVP_aes_192_gcm}, 24, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+    {{"aes_gcm"},     {&EVP_aes_256_gcm}, 32, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+    {{"aes_128_gcm"}, {&EVP_aes_128_gcm}, 16, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+    {{"aes_192_gcm"}, {&EVP_aes_192_gcm}, 24, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+    {{"aes_256_gcm"}, {&EVP_aes_256_gcm}, 32, AEAD_CIPHER, {{EVP_CTRL_GCM_SET_IVLEN,EVP_CTRL_GCM_GET_TAG,EVP_CTRL_GCM_SET_TAG}}},
+#else
+    {{"aes_gcm"}, {NULL}, 0, AEAD_CIPHER, {{0,0,0}}},
```

Comment:
I'm not following why `aes_ctr` is treated differently than `aes_gcm` or `aes_ccm`.

It seems like the current code wants to have definitions for everything so that you can raise exception `badarg` if it's an unexpected algorithm.  If it's unsupported because `cipher.p` is NULL, then you raise exception `atom_notsup` instead.

If that's the case, shouldn't there be `aes_128_gcm` etc in the case of `!HAVE_GCM`?